src/hotspot/cpu/riscv/macroAssembler_riscv.cpp

```diff
@@ -2425,7 +2429,20 @@ int MacroAssembler::corrected_idivq(Register result, Register rs1, Register rs2,
 
   int idivq_offset = offset();
   if (!want_remainder) {
-    div(result, rs1, rs2);
+    if (is_signed) {
+      div(result, rs1, rs2);
+    } else {
+      Label Lltz, Ldone;
+      bltz(rs2, Lltz);
```

Comment:
I am not quite sure what this `bltz` branch is for. Is this a minor performance tunning here? And How would this make a difference then if that's true? I didn't see much difference from the LongDivMod.testDivideUnsigned `negative` jmh test result.

Reply:
+1. It's also the only test case where there is a regression on the JMH numbers, or at least not a clear improvement (before: 6385.280, after: 6433.223)

On your JMH numbers, how many iterations have you run for each benchmark? I don't see the standard deviation which would be useful to better understand noise.

Reply:
`For the algorithm details, check j.l.Long::divideUnsigned` in the jdk lib source, it mentions this algorithm.

It's not related to the difference between negative and positive test cases, it's related to the cost of divxx instructions, compared to the lines between 2440 ~ 2443 in src/hotspot/cpu/riscv/macroAssembler_riscv.cpp, the divu cost for negative value is still very high.

```
int_def ALU_COST             (  100,  1 * DEFAULT_COST);
int_def BRANCH_COST          (  200,  2 * DEFAULT_COST);
int_def IDIVDI_COST          ( 6600, 66 * DEFAULT_COST);
```

I have also re-run the benchmark with more warmup (5) and iteration (10), please check the data in pr desc.
I also attach the diff between v1 and v2 intrinsic. v2 is this patch. v1 is diff based on v2, it just use riscv divxx directly without optimization for negative value brong by the algorithm.

Reply:
I don't know why the previous jmh data has no `error` part, maybe because it's too low to show.

Reply:
IIUC, with the branch, the results are `6376.674 ±  16.869  ns/op`, and without the branch, they are `29518.033 ± 49.056`, correct? If so, the branch makes more sense, at least of the board you've tested.

Reply:
Yes, for negtive divisor, unsigned div can go through a quick path which is must faster than built-in instructions.
And this is demonstrated in the div cost in riscv.ad, and also verified by benchmark tests run on the board.

( I'm not sure if in the future built-in div will be faster, if it turns out in the future, we should also need to redefine the div cost in riscv.ad, and re-visit this intrinsic. )

Reply:
Ok, so let's keep the version with branch. You should add a comment at https://github.com/openjdk/jdk/pull/16346/files#diff-7a5c3ed05b6f3f06ed1c59f5fc2a14ec566a6a5bd1d09606115767daa99115bdR2435 explaining just that.

Reply:
But you're putting a conditional branch in the way of the common cases, and you're greatly increasing icache pressure, for the sake of a rare case. How does that make any sense?

Reply:
I agree with @theRealAph here, this seems like a micro optimisation for the sake of microbenchmark that will not be beneficial in general. And if a considerable portion of divisors does lie in this range, the optimisation can always be applied from the caller side. Furthermore, by doing so we will even have the benefit of branch profiling, which will help achieve better results. Another note is that I do not know any compiler that does this premature optimisation. Thanks.

Reply:
Thanks @theRealAph @merykitty for the comments, I agree with you.
Will use the divu instead of introducing a cond branch here. And we will consider the regression of negative (also mixed) test cases as rare case, so just make the common case (positve one) get optimized.